servers/rendering/rendering_shader_container.h

```diff
@@ -120,38 +124,167 @@ class RenderingShaderContainer : public RefCounted {
 	virtual uint32_t _to_bytes_shader_extra_data(uint8_t *p_bytes, uint32_t p_index) const;
 	virtual uint32_t _to_bytes_footer_extra_data(uint8_t *p_bytes) const;
 
-	// This method will be called when set_from_shader_reflection() is finished. Used to update internal structures to match the reflection if necessary.
-	virtual void _set_from_shader_reflection_post(const RenderingDeviceCommons::ShaderReflection &p_reflection);
+	template <class T>
+	struct ReflectSymbol {
+		static constexpr uint32_t STAGE_INDEX[RDC::SHADER_STAGE_MAX] = {
+			0, // SHADER_STAGE_VERTEX
+			1, // SHADER_STAGE_FRAGMENT
+			0, // SHADER_STAGE_TESSELATION_CONTROL
+			1, // SHADER_STAGE_TESSELATION_EVALUATION
+			0, // SHADER_STAGE_COMPUTE
+		};
+
+		BitField<RDC::ShaderStage> stages = {};
+
+	private:
+		const T *_spv_reflect[2] = { nullptr };
+
+	public:
+		_FORCE_INLINE_ constexpr uint32_t get_index_for_stage(RDC::ShaderStage p_stage) const {
+			DEV_ASSERT(stages.has_flag((1 << p_stage)));
+			return STAGE_INDEX[p_stage];
+		}
+
+		const T &get_spv_reflect(RDC::ShaderStage p_stage) const;
+
+		/*! Returns the first valid stage if multiple stages are set.
+		 *
+		 * Crashes if no stages are set.
+		 */
+		const T &get_spv_reflect() const {
+			for (const T *d : _spv_reflect) {
+				if (d != nullptr) {
+					return *d;
+				}
+			}
+			CRASH_NOW_MSG("No stages set in ReflectSymbol");
+		}
+		void set_spv_reflect(RDC::ShaderStage p_stage, const T *p_spv);
+	};
+
+	struct ReflectImageTraits {
+		RDC::DataFormat format = RDC::DATA_FORMAT_MAX;
+	};
+
+	struct ReflectUniform : ReflectSymbol<SpvReflectDescriptorBinding> {
+		RDC::UniformType type = RDC::UniformType::UNIFORM_TYPE_MAX;
+		uint32_t binding = 0;
+
+		ReflectImageTraits image;
+
+		uint32_t length = 0; // Size of arrays (in total elements), or ubos (in bytes * total elements).
+		bool writable = false;
+
+		bool operator<(const ReflectUniform &p_other) const {
+			if (binding != p_other.binding) {
+				return binding < p_other.binding;
+			}
+			if (type != p_other.type) {
+				return type < p_other.type;
+			}
+			if (writable != p_other.writable) {
+				return writable < p_other.writable;
+			}
+			if (stages != p_other.stages) {
+				return stages < p_other.stages;
+			}
+			if (length != p_other.length) {
+				return length < p_other.length;
+			}
+			return false;
+		}
+	};
 
-	class ReflectedShaderStage {
+	struct ReflectSpecializationConstant : ReflectSymbol<SpvReflectSpecializationConstant> {
```

Comment:
I'm not fully understanding the purpose of duplicating ShaderSpecializationConstant here. After reading through the code in the CPP file I am left feeling like most of the changes here are superficial and have the unintended consequence of duplicating the API without a corresponding gain.